### PR TITLE
Add golangci-lint to Github CI step

### PR DESCRIPTION
**Pull request by stolsma**
@antoninbas First stab it implementing golangci-lint. Just basic use of the action but the CI run on this PR on my fork already gave a lot of errors (see the list below). We need to find out which options (defined in .golangci.yml) are reasonable and which not. There are also a lot of other options related to errors showing up in the checked PR I do not understand yet. Will play with it a little longer.

Do not merge as I put my local PR branch in the go.yml... 

*Current Error list:* 
```
  Error: Error return value of `stream.CloseSend` is not checked (errcheck)
  Error: Error return value of `stream.Send` is not checked (errcheck)
  Error: Error return value of `stream.Send` is not checked (errcheck)
  Error: Error return value of `p4RtC.Run` is not checked (errcheck)
  Error: Error return value of `p4RtC.Run` is not checked (errcheck)
  Error: func IpToBinary should be IPToBinary (golint)
  Error: type name will be used as client.ClientOptions by other packages, and that stutters; consider calling this Options (golint)
  Error: error strings should not be capitalized or end with punctuation or a newline (golint)
  Error: error strings should not be capitalized or end with punctuation or a newline (golint)
  Error: method tableId should be tableID (golint)
  Error: method actionId should be actionID (golint)
  Error: method digestId should be digestID (golint)
  Error: method counterId should be counterID (golint)
  Error: `if` block ends with a `return` statement, so drop this `else` and outdent its block (golint)
  Error: line is 133 characters (lll)
  Error: struct of size 64 bytes could be of size 56 bytes (maligned)
  Error: S1021: should merge variable declaration with assignment on next line (gosimple)
  Error: S1021: should merge variable declaration with assignment on next line (gosimple)
  Error: composites: `github.com/antoninbas/p4runtime-go-client/pkg/client.ExactMatch` composite literal uses unkeyed fields (govet)
  Error: composites: `github.com/antoninbas/p4runtime-go-client/pkg/client.ExactMatch` composite literal uses unkeyed fields (govet)
  Error: composites: `github.com/antoninbas/p4runtime-go-client/pkg/client.ExactMatch` composite literal uses unkeyed fields (govet)
  Error: unreachable: unreachable code (govet)
  Error: SA1019: package github.com/golang/protobuf/proto is deprecated: Use the "google.golang.org/protobuf/proto" package instead.  (staticcheck)
  Error: SA4017: Errorf is a pure function but its return value is ignored (staticcheck)
  Error: SA5011: possible nil pointer dereference (staticcheck)
```

### CODE REVIEW

**Review by antoninbas — COMMENTED** [review]
Thanks for working on this. I do feel we should provide a mechanism for people to run this locally as well, through a `Makefile` target that install and runs the correct version of `golangci-lint`.

I did the same thing a while back for a different project, if this is helpful: https://github.com/vmware-tanzu/antrea/blob/fbbda2be35dd453c16e9286e685a9faba2424c77/Makefile#L234-L250

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -36,3 +37,18 @@ jobs:
...
+    - uses: actions/checkout@v2
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v2
+      with:
+        # Required: the version of golangci-lint is required and must be specified without patch version:
+        # we always use the latest patch version.
+        version: v1.38
+
+        # Optional: show only new issues if it's a pull request. The default value is `false`.
+        only-new-issues: true
```
[the remark this answers is not in the document]
Comment (stolsma) [reply .github/workflows/go.yml:54]: Agreed!

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
+linters:
```
[the remark this answers is not in the document]
Comment (stolsma) [reply .golangci.yml:1]: Nice! Will copy relevant parts!

**Comment by stolsma** [comment]
Agree to the local use script, was considering it earlier but forgot to add it...

**Comment by stolsma** [comment]
@antoninbas Implemented all the requested changes and added 2nd commit with fixes on the current lint errors. Still have two errors remaining.

```
  Error: SA1019: package github.com/golang/protobuf/proto is deprecated: Use the "google.golang.org/protobuf/proto" package instead.  (staticcheck)
  Error: SA5011: possible nil pointer dereference (staticcheck)
  Error: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
```  
1. Tried to solve the deprecated proto package but don't know how.. :-( Maybe you could have a look at it?
2. I'm looking into the nil pointer dereference errors as I don't understand why the linter doesn't suppress because of the added lint directive...

Any more things that need to be done??

**Comment by stolsma** [comment]
Ping? @antoninbas (no hurry, just curious about your opinion...)

File: pkg/client/tables.go [hunk 3]
```diff
@@ -220,6 +220,7 @@ func (c *Client) NewTableEntry(
 		Action:          action,
 	}
 
+	//lint:ignore SA5011 if mfs==nil then for loop is not executed by default
```
Comment (antoninbas) [inline pkg/client/tables.go:223]: looks like your comment to disable the false-positive warning is not right as we still get the failure?

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -36,3 +36,18 @@ jobs:
...
+    name: lint
+    runs-on: [ubuntu-latest]
+    steps:
+    - uses: actions/checkout@v2
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v2
+      with:
+        # Required: the version of golangci-lint is required and must be specified without patch version:
+        # we always use the latest patch version.
+        version: v1.39
```
Comment (antoninbas) [inline .github/workflows/go.yml:50]: maybe add a comment that this needs to be kept in sync with the version in the Makefile

File: .github/workflows/go.yml [hunk 5]
```diff
@@ -36,3 +36,18 @@ jobs:
...
+
+        # Optional: show only new issues if it's a pull request. The default value is `false`.
+        only-new-issues: false
```
Comment (antoninbas) [inline .github/workflows/go.yml:53]: can you add back the missing newline?

File: .golangci.yml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  skip-dirs-use-default: true
+
+linters-settings:
+  goimports:
+    local-prefixes: github.com/antoninbas/p4runtime-go-client/
+
+linters:
+  disable-all: true
+  enable: # see https://golangci-lint.run/usage/linters/
+    #
```
Comment (antoninbas) [inline .golangci.yml:15]: I don't think I see a reason to provide the full list of all the linters commented out. I think we should just keep the ones we use. You already have the link above (`https://golangci-lint.run/usage/linters/`) if someone needs to enable a new one.

Comment (stolsma) [reply .golangci.yml:15]: Ack!

File: .github/workflows/go.yml (hunk 5 above)
Comment (stolsma) [reply .github/workflows/go.yml:53]: Ack!

File: pkg/client/tables.go (hunk 3 above)
Comment (stolsma) [reply pkg/client/tables.go:223]: I still don't understand why this line is wrong. Tried everything! The VSC lint implementation takes the line and removes the lint error but golangci-lint doesn't take it and gives error.... Archhh... :-(

**Review by antoninbas — APPROVED** [review]

File: .golangci.yml (hunk 6 above)
Comment (rulisastra) [reply .golangci.yml]: > Ack!

LOL. Your same replies got me laughing in the middle of the night

**Pull request merged**